[QUANT] Add GPTQModel Dynamic Quantization + `lm_head` Quantization

Qubitium opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

Per module GPTQ quantization control/override + Allow loading of quantized `lm_head`:

###  Before PR: 
- All GPTQ `modules` share the same `GPTQConfig` (bits, group_size, etc)
-  Cannot load quantized `lm_head`

### This PR: 

- Every `module` can have unique `GPTQConfig` (bits, group_size, etc)
- Specific `modules` can be optionally skipped entrirely for quantization load based on `dynamic` override. 
- Quantized `lm_head` can be loaded

`Dynamic` Logic:

- If rule match, matching with `prefix` aka module full path name such as `mlp.down_proj`.
- if `negative` rule match, the module is skipped and loaded as normal, non-gptq quantized module
- if `positive` rule match, gptq config for this module is overriden by the key/value dicated in the match rule.  For example, you can optionally override the `bits`, `group_size` per module. 
- If `None` or no-match, nothing happens. Base `GPTQConfig` is used. 

Notes:

* Depends on GPTQ models quantized by [GPTQModel](https://github.com/modelcloud/gptqmodel)
* Backward compatible with all existing non-dynamic GPTQ models. 

`dynamic` config sample/doc: https://github.com/ModelCloud/GPTQModel?tab=readme-ov-file#dynamic-quantization-per-module-quantizeconfig-override

## Modifications

* `dynamic` property added to `quantization_config.json` in gptq models which contains regex rules paired with overriding values in dict format.  The actual code for this feature is in vLLM (0.7.3). The dynamic override logic code is directly copied from vllm, also written by us.

* `prefix` (module weight path) must be passed down in the loading logic so `dynamic` override can take effect before GPTQ linear layer creation.  The `dynamic` override rules uses the `prefix` (weight path/name) value for matching and if match found, `GPTQConfig` base properties are overriden. 
* `lm_head.linear_method` property changed to `lm_head.quant_method` to allow correct GPTQ linear layer loading


TODO:

* [x] Validate Working State
* [x] Add `dynamic` GPTQ model unit tests
* [x]  Clean-up code/structure

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [x] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

ZX-ModelCloud committed:
Changed VocabParallelEmbedding.linear_method to quant_method to be consistent with LinearBase.

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
call param.packed_factor instead of param.pack_factor

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
add monkey_patch_vllm_get_linear_quant_method()

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
pass prefix argument

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
fix gptq_marlin error

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
cleanup

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
add prefix

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
add prefix

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

Qubitium committed:
use clearer api name and re-order args

Qubitium committed:
format

Qubitium committed:
move import to top

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium committed:
reduce vllm depend: move dynamic config extraction method to sglang

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

ZX-ModelCloud committed:
add unittest

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
update unittest

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

ZX-ModelCloud committed:
code format

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

Qubitium commented:
@merrymercy @Ying1123 @zhyncs Ready for code review. Plesse trigger the sglang CI so we can be sure there are no regressions. 

The code/logic follows our vLLM code of same core feature released `v0.7.3`.

- The biggest changes to SGLang is our maintaining and passing `prefix` value in the loading code. `Prefix` is the module weight full path/key name in the weight file. Even though GPTQModel dynamic is the only one using this `prefix` but we believe it will be widely used in other qunatization frameworks `dynamic` feature is they choose to add them. 
- The other part of the `overrdie_` method code we copied our code from vLLM and duplicated here so there is less dependency on specific version of vLLM.

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium commented:
@zhaochenyang20 Tagging you to see if we can get this PR review started. We modified many `core` `model` files but the changes are minimal per file and reasons for `prefix` weight path push down is explained. Thanks!

zhaochenyang20 commented:
Thanks @yinfan98 @BBuf Could please take a look?

zhaochenyang20 reviewed (changes requested):
hey, please add the unit test in the run suite files. 

https://github.com/sgl-project/sglang/blob/main/test/README.md

Thanks so much

zhaochenyang20 commented:
@Qubitium Also, rebase with the main plz

Edenzzzz commented on python/sglang/srt/models/deepseek_v2.py:
```diff
@@ -241,43 +255,47 @@ def __init__(
             self.hidden_size,
             bias=False,
             quant_config=quant_config,
+            prefix=f"{prefix}.o_proj",
```
Would it be possible to not add a prefix for every model, just use PyTorch's `named_parameters()` dict? It will show layers exactly as their variable names

Qubitium replied:
@Edenzzzz  Problem is that in the inner submodules kernel (linear_layer) selection stage, it doesn't know who the parent is so it can't generate that info (the module fulll path name as it exists in the weight file) and it must be passed down from parent module to child module. The decision for quant kernel selection is made at the lower module init level not at the top where `named_parameters` would hold the complete tree.

Edenzzzz replied:
I see, that's because you pass in a single low-level layer. Thanks for clarifying.

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium committed:
add gptqmodel tests to run_suite.py

Qubitium commented:
> hey, please add the unit test in the run suite files.
> 
> https://github.com/sgl-project/sglang/blob/main/test/README.md
> 
> Thanks so much

CI tests added to `run_suites.py`. Please trigger. Thanks!

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

zhaochenyang20 commented:
@Qubitium Thanks so much! We needs more reviewers on this.

zhaochenyang20 commented:
https://github.com/sgl-project/sglang/blob/main/docs/backend/quantization.md

Could you also update this docs plz?

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium committed:
Update quantization.md

Qubitium commented:
> https://github.com/sgl-project/sglang/blob/main/docs/backend/quantization.md
> 
> Could you also update this docs plz?

@zhaochenyang20  I updated the quantization doc to include GPTQModel, added sample for offline quantization, and also cleaned up the doc as a whole. Please check.

zhaochenyang20 commented:
@Qubitium Thanks. BBuf will help to review @BBuf

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

zhaochenyang20 commented:
@Qubitium fix the lint and rebase please

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium committed:
format

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium commented:
> hey, please add the unit test in the run suite files.
> 
> https://github.com/sgl-project/sglang/blob/main/test/README.md
> 
> Thanks so much

@zhaochenyang20  I still see the PR info that there is a requested change from you but I can't view any changes. Please double check.

zhaochenyang20 commented:
https://github.com/sgl-project/sglang/actions/runs/13615599542/job/38058236758?pr=3790

@Qubitium This test failes. Please solve it locally.

Qubitium commented:
> https://github.com/sgl-project/sglang/actions/runs/13615599542/job/38058236758?pr=3790
> 
> @Qubitium This test failes. Please solve it locally.

This PR require `GPTQConfig` and `MarlinConfig` from vllm `v0.7.3` but I don't think SGLang depends has caught up to vLLM 0.7.3. 

I will try to move these 2 files over so GPTQ code is less dependent on vllm.

zhaochenyang20 commented:
> > https://github.com/sgl-project/sglang/actions/runs/13615599542/job/38058236758?pr=3790
> > @Qubitium This test failes. Please solve it locally.
> 
> This PR requires `GPTQConfig` and `MarlinConfig` from vllm `v0.7.3` but I don't think SGLang depends has caught up to vLLM 0.7.3.
> 
> I will try to move these 2 files over so GPTQ code is less dependent on vllm.

Thanks. We should remove dependency from VLLM.

Qubitium committed:
remove vllm depends

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium committed:
remove more vllm 0.7.3 specific depend

Qubitium committed:
Merge branch 'compat_gptqmodel_dynamic' of https://github.com/ZX-ModelCloud/sglang into compat_gptqmodel_dynamic

zhaochenyang20 commented:
@zhyncs 

Not sure could we add this kind of dependency?

python/sglang/srt/layers/quantization/gptq.py

Qubitium commented:
> @zhyncs
> 
> Not sure could we add this kind of vllm dependency?
> 
> python/sglang/srt/layers/quantization/gptq.py

@zhaochenyang20 The depends are not more, but less. The `python/sglang/srt/layers/quantization/gptq.py` file appears to add more vllm depends but that's because they are already imported dynamically when going through the GPTQ and Marlin kernel related loading code. For `python/sglang/srt/layers/quantization/gptq.py` I only imported the necessary `Config` classes which I believe is safe and removes any type of GPTQ `Config` dependency on vllm 0.7.3. Otherwise, this PR cannot run.

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

sleepcoo commented:
Can you add some unit tests to ensure that future iterations don't impact the GPTQ functionality? cc@zhaochenyang20

Qubitium committed:
all prefix code use add_prefix

Qubitium committed:
Merge branch 'compat_gptqmodel_dynamic' of https://github.com/ZX-ModelCloud/sglang into compat_gptqmodel_dynamic

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium committed:
format

Qubitium commented:
Update TODO

* [ ] There is a `GPTQ` kernel regression in our CI test that may be upstream (vLLM). Marlin kernel has no issue.  We are checking. 

@sleepcoo Which CI tests do you have in mind?

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium committed:
simplify

ZX-ModelCloud committed:
assert output

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

yinfan98 commented:
Great work, there are some vllm dependencies that need to be removed.

Qubitium commented on python/sglang/srt/layers/quantization/gptq.py:
```diff
@@ -0,0 +1,412 @@
+import logging
+from fractions import Fraction
+from typing import Any, Dict, List, Optional, Union
+
+import torch
+from vllm.model_executor.layers.quantization.gptq import GPTQLinearMethod
+from vllm.model_executor.layers.quantization.gptq_marlin import (
+    GPTQMarlinLinearMethod,
+    GPTQMarlinMoEMethod,
+)
+from vllm.model_executor.layers.quantization.marlin import MarlinLinearMethod
+from vllm.model_executor.layers.quantization.utils.marlin_utils import (
+    check_marlin_supported,
+)
+from vllm.platforms import current_platform
+from vllm.scalar_type import scalar_types
```
@yinfan98  These are not `extra` vllm depends.  SGLang already imports them except indirectly via calls to vLLM GPTQ/Marlin code paths. Meaning, they are always imported when GPTQ are invocked. Here, I am migrating the GPTQ related Config classes here so there is no dependency on vllm 0.7.3. SGlang only synced to vllm 0.7.2.

Qubitium commented:
> I find it quite cumbersome to change dozens of model structure files just because parameter names have been modified. Would it be possible to provide a script that restores the parameter names exported by GPTQ to their original format? This way, the current PR would be much cleaner, and it would require fewer changes to the existing code. @Qubitium

Good question and I will address the issue here since you are not will not be the last person to ask this valid question.

1. GPTQModel quantized GPTQ models does not produce non-compatible weight paths. Meaning, the weight paths in the weight files are normal, non-modified, weight paths identical to original model except for the extra suffix gptq only weights like `mlp.down_proj.[suffix]` where suffix in `[qweight, scale, qzero, g_idx]`

2. GPTQ `dynamic` quantization must happen at the Marlin (GPTQ) nn.module kernel (mostly LinearLayer) pre-init stage since need the overrides to happen: base 4bit gptq now overridden to 8bit, so pass 8bits to Marlin init for example.

3.  The model loading happens from TOP to DOWN and dictated by `modeling` files, thus why all the changes. Not BOTTOM to UP.  So for example Llama model has a `LllamaDecodeLayer` which are all the repeating layers. The modeling file creates (inits) all the correct sub-modules, their correct parameters names and config (critical here as all the config including quantization/GPTQ must be correct pre-init), in a drill-down tree fasion. Once the modeling file has correctly produced all the correct modules. The weights are loaded to the same named `parameters`of these `base` submodules. Some modules like `LlamaDecodeLayer` are just logic holders to real `base` sub-modules or holders.

4. Chicken and egg issue. The lowest, primal, or I call `base` nn.Modules, is where all the weight/matmul happen and where the GPTQ linear layer (aka kernel) are inited. At this level, the `base` module has no idea where it is or who it is: name or `prefix` is needed to give this module an full-id identification. After all the `prefix` passing from top to bottom (parent to children), we have an `module id` aka full `prefix` of `layers.0.mlp.down_proj` for example. Without passing down the `prefix`, sub-module doesn't even know who it is other than it's own instance type. 

5. Before GPTQ kernel layers are created, we do `dynamic` matching of the `base` module  `id` == `prefix` and if matches, perform the correct property overrrides. 

I would love to have another way to do this. Manually adding prefix to every single nn.module in every modeling file is not my idea of fun either. =)

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

zhaochenyang20 commented:
@Qubitium should be rebased

zhaochenyang20 commented:
Thanks so much. After remove the dependency, we shall merge it.

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium commented:
> Thanks so much. After remove the dependency, we shall merge it.

Rebased. Waiting for CI test to run. I am still confused at what dependency that I should remove? Right now, the code is compatbile with vllm 0.7.2. I have already removed the dependencies which require vllm 0.7.3.

Qubitium committed:
fix ci

zhaochenyang20 commented:
@Qubitium still have some conflicts. I will ask @yizhang2077 to help on review, "how to remove vllm dependency"

Qubitium commented:
> @Qubitium still have some conflicts. I will ask @yizhang2077 to help on review, "how to remove vllm dependency"

@zhaochenyang20  @yizhang2077  I already addressed the fake increase of vllm dependency in this reply:
https://github.com/sgl-project/sglang/pull/3790#issuecomment-2693068292

There is no increase of vLLM dependency. Zero. Actually, there is net negative, less vLLM dependency since we moved a few GPTQ config classes inside SGLang. On first glance of the gptq.py changes, you might think the below are extra depends:

```py
import logging
from fractions import Fraction
from typing import Any, Dict, List, Optional, Union

import torch
from vllm.model_executor.layers.quantization.gptq import GPTQLinearMethod
from vllm.model_executor.layers.quantization.gptq_marlin import (
    GPTQMarlinLinearMethod,
    GPTQMarlinMoEMethod,
)
from vllm.model_executor.layers.quantization.marlin import MarlinLinearMethod
from vllm.model_executor.layers.quantization.utils.marlin_utils import (
    check_marlin_supported,
)
from vllm.platforms import current_platform
from vllm.scalar_type import scalar_types
```

But all the above are already implicitt depends when SGLang run any GPTQ related kernel since SGLangs calls vLLM to load it's GPTQ kernels. What I mean is that, they are imported 100% of the time if you use quantization with GPTQ. Here, I am just calling them all expliciity. Before this PR, the imports are `hidden from view`, this PR exposes them. 

In execution, this PR actually reduced vLLM dependency, not more.

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium commented:
Latest CI error appears unrelated to our changes. `KeyError: 'debug_tensor_dump_output_folder'`. I have no idea what this variable is within SGLang. Is this injected by CI?

[KeyError: 'debug_tensor_dump_output_folder'](https://github.com/sgl-project/sglang/actions/runs/13643964727/job/38139337684?pr=3790#step:4:273)

zhaochenyang20 commented:
https://github.com/sgl-project/sglang/actions/runs/13643964727/job/38139338876?pr=3790

@Qubitium Could you fix this?

zhaochenyang20 commented:
Also, thanks so much for the PR. You are MY HERO. Too big PR to review and to do 😂 @Qubitium

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

zhaochenyang20 commented:
@Qubitium I will merge it after the CI. thanks!

Qubitium committed:
try to fix circular imports from vllm

Qubitium committed:
try (2): fix circular imports

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium commented:
> @Qubitium I will merge it after the CI. thanks!

The merge/conflict with `main` is never ending! lol. God I hope the conflicts stop happenings since we touched every single file. I guess we brought this on ourselves. Crying. lol

Qubitium committed:
format

Qubitium commented:
@zhaochenyang20 Previous failed tests passing! Remaining 2 failed tests appears unrelated to PR.

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium committed:
Merge branch 'main' into compat_gptqmodel_dynamic

Qubitium commented:
@zhaochenyang20 Can you check? All the CI tests are good. The only test failing appears unrelated. I don't want to do more rebase with master because it's clean and by the time you review the CI, the master is out of sycn again (but clean mergeable) so shold have no issues. If there is any conflict, I will merge but want to avoid merging since is soo much merging activity.

zhaochenyang20 commented:
@Qubitium will merge it right now
